pimd/pim_register.c

```diff
@@ -131,31 +153,40 @@ int pim_register_stop_recv(struct interface *ifp, uint8_t *buf, int buf_size)
 	pim_parse_addr_ucast(&source, buf, buf_size);
 	sg.src = source.u.prefix4;
 
-	upstream = pim_upstream_find(pim, &sg);
-	if (!upstream) {
-		return 0;
-	}
 
 	if (PIM_DEBUG_PIM_REG)
-		zlog_debug("Received Register stop for %s", upstream->sg_str);
+		zlog_debug("Received Register stop for %s",
+			   pim_str_sg_dump(&sg));
+
+	rp = RP(pim_ifp->pim, sg.grp);
+
+	if (sg.src.s_addr == rp->rpf_addr.u.prefix4.s_addr) {
+		sg.src.s_addr = INADDR_ANY;
+	} else {
+		up = pim_upstream_find(pim, &sg);
+		if (up)
+			pim_reg_stop_upstream(pim, up);
 
-	switch (upstream->reg_state) {
-	case PIM_REG_NOINFO:
-	case PIM_REG_PRUNE:
-		return 0;
-	case PIM_REG_JOIN:
-		upstream->reg_state = PIM_REG_PRUNE;
-		pim_channel_del_oif(upstream->channel_oil, pim->regiface,
-				    PIM_OIF_FLAG_PROTO_PIM, __func__);
-		pim_upstream_start_register_stop_timer(upstream, 0);
-		pim_vxlan_update_sg_reg_state(pim, upstream,
-			false/*reg_join*/);
-		break;
-	case PIM_REG_JOIN_PENDING:
-		upstream->reg_state = PIM_REG_PRUNE;
-		pim_upstream_start_register_stop_timer(upstream, 0);
 		return 0;
 	}
+	/*
```

Comment:
I'd rather not search the entire S,G upstream space if we happen to have a (*,G) on the box.

Reply:
Fixed it.